src/eventstream-proxy/eventstream-proxy.base.ts

```diff
@@ -66,55 +68,86 @@ export abstract class EventStreamProxyBase extends WebSocketEventsBase {
-      Promise.all(this.connectListeners.map(l => l.onConnect()))
-        .then(() => {
-          this.setCurrentClient(client);
-          this.startListening();
-        })
-        .catch(err => {
-          this.logger.error(`Error initializing event stream proxy: ${err}`);
-        });
-    }
+    client.on('message', async (message: string) => {
+      const action = JSON.parse(message) as WebSocketActionBase;
+      switch (action.type) {
+        case 'start':
+          const startAction = action as WebSocketStart;
+          this.startListening(client, startAction.namespace);
+          break;
+      }
+    });
   }
 
   private queueTask(task: () => void) {
     this.queue = this.queue.finally(task);
   }
 
-  private startListening() {
+  private async startListening(client: WebSocketEx, namespace: string) {
     if (this.url === undefined || this.topic === undefined) {
       return;
     }
-    this.socket = this.eventstream.connect(
-      this.url,
-      this.topic,
-      events => {
-        this.queueTask(() => this.processEvents(events));
-      },
-      receipt => {
-        this.broadcast('receipt', <EventStreamReply>receipt);
-      },
-    );
+    try {
+      if (!this.namespaceEventStreamSocket.has(namespace)) {
+        const eventStreamSocket = await this.eventStreamService.connect(
+          this.url,
+          this.topic,
+          namespace,
+          events => {
+            this.queueTask(() => this.processEvents(events, namespace));
+          },
+          receipt => {
+            this.broadcast('receipt', <EventStreamReply>receipt);
+          },
+        );
+        this.namespaceEventStreamSocket.set(namespace, eventStreamSocket);
+      }
+      let clientSet = this.namespaceClients.get(namespace);
+      if (!clientSet) {
+        clientSet = new Set<WebSocketEx>();
+      }
+      clientSet.add(client);
+      this.namespaceClients.set(namespace, clientSet);
+
+      // ack the start command
+      client.send(
+        JSON.stringify({
+          event: 'started',
+          data: {
+            namespace: namespace,
+          },
+        }),
+      );
+      this.logger.debug(`Started namespace '${namespace}'`);
+    } catch (e) {
+      this.logger.error(`Error connecting to event stream websocket: ${e.message}`);
+    }
   }
 
   handleDisconnect(client: WebSocketEx) {
     super.handleDisconnect(client);
-    if (this.server.clients.size === 0) {
-      this.stopListening();
-    } else if (client.id === this.currentClient?.id) {
-      for (const newClient of this.server.clients) {
-        this.setCurrentClient(newClient as WebSocketEx);
-        break;
-      }
-    }
-  }
 
-  private stopListening() {
-    this.socket?.close();
-    this.socket = undefined;
-    this.currentClient = undefined;
+    // Iterate over all the namespaces this client was subscribed to
+    this.namespaceClients.forEach((clientSet, namespace) => {
+      clientSet.delete(client);
```

Comment:
Shouldn't we be checking the return value to see if this set included this client? Seems like the current behavior will nack all messages on all namespaces, regardless of whether the disconnect client was actually subscribed to them.